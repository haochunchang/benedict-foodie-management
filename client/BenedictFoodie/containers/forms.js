import React, { useState } from 'react';
import {
    View, Text, TextInput, Button, Alert,
} from "react-native"
import { Select, Option } from '../third_party/react-native-select-list-modified/src';

const SatisfactionScoreDescription = {
    1: "Does not eat with snack added",
    2: "Finish eating with snack",
    3: "Finish eating",
    4: "Eating eagerly",
    5: "Eating eagerly with talking",
}

const HEADERS = {
    Accept: 'application/json',
    'Content-Type': 'application/json'
};

const FoodTypeDropdownList = ({ handle }) => {
    return (
        <Select onSelect={(value) => { handle(value) }}>
            <Option value={'dry'}>Dry Food</Option>
            <Option value={'wet'}>Wet Food</Option>
            <Option value={'snack'}>Snack</Option>
        </Select >
    );
};

const SatisfactionScoreDropdown = ({ initScore, handle }) => {
    return (
        <Select default={SatisfactionScoreDescription[initScore]} onSelect={(value) => { handle(value) }}>
            <Option value={1}>{SatisfactionScoreDescription[1]}</Option>
            <Option value={2}>{SatisfactionScoreDescription[2]}</Option>
            <Option value={3}>{SatisfactionScoreDescription[3]}</Option>
            <Option value={4}>{SatisfactionScoreDescription[4]}</Option>
            <Option value={5}>{SatisfactionScoreDescription[5]}</Option>
        </Select >
    );
};

export const FoodForm = ({ closeForm, backendUrl }) => {
    const [name, onChangeName] = useState("");
    const [type, onChangeType] = useState("");
    const [desc, onChangeDesc] = useState("");
    const [isLoading, onChangeIsLoading] = useState(false);

    const submitFoodForm = () => {
        const food = {
            Name: name,
            Type: type,
            Description: desc,
        };
        onChangeIsLoading(true);
        fetch(`${backendUrl}/foods`, {
            method: 'POST',
            headers: HEADERS,
            body: JSON.stringify(food)
        })
            .then((resp) => resp.json())
            .then((data) => {
                if (data.message != "Food created") {
                    throw data.message;
                }
                onChangeIsLoading(false);
                closeForm();
            }).catch((error) => {
                Alert.alert("Error", error, [{ text: "Okay" }]);
                onChangeIsLoading(false);
            });
    }

    return (
        <View>
            <Text>Creat Food Stock</Text>
            <Text>Food name</Text>
            <TextInput
                onChangeText={onChangeName}
                value={name}
                placeholder="Enter the food name"
                autoFocus={true}
            />
            <Text>Food type</Text>
            <FoodTypeDropdownList handle={onChangeType} />
            <Text>Food description</Text>
            <TextInput
                onChangeText={onChangeDesc}
                value={desc}
                placeholder="What's about the food?"
            />
            <Button title="Submit" onPress={submitFoodForm} disabled={isLoading} />
            <Button title="Cancel" onPress={closeForm} />
        </View>
    )
}

export const RecordForm = ({ record, closeForm, backendUrl, thisMonthRecord, onChangeThisMonthRecord }) => {

    const eatingDate = record.EatingDate;
    const [name, onChangeName] = useState(record.Food.Name);
    const [quantity, onChangeQuantity] = useState(record.EatenQuantity);
    const [score, onChangeScore] = useState(record.SatisfactionScore);
    const [desc, onChangeDesc] = useState(record.Description);
    const [isLoading, onChangeIsLoading] = useState(false);
    // TODO: handle images uploading
    // const [photoURL, onChangePhotoURL] = useState("");

    const submitRecord = () => {
        const record = {
            Food: { Name: name },
            EatingDate: eatingDate,
            EatenQuantity: Number.parseFloat(quantity),
            SatisfactionScore: score,
            Description: desc
        };
        onChangeIsLoading(true);
        fetch(`${backendUrl}/records`, {
            method: 'POST',
            headers: HEADERS,
            body: JSON.stringify(record)
        })
            .then((resp) => resp.json())
            .then((_) => {
                const d = thisMonthRecord;
                d[record.EatingDate] = record;
                d[record.EatingDate].isModifying = true;
                onChangeThisMonthRecord(d);
                onChangeIsLoading(false);
                closeForm();
            }).catch((error) => {
                Alert.alert("Error", error.message, [{ text: "Okay" }]);
                onChangeIsLoading(false);
            });
    };

    const updateRecord = () => {
        const newRecord = {
            Food: { Name: name },
            EatingDate: eatingDate,
            EatenQuantity: Number.parseFloat(quantity),
            SatisfactionScore: score,
            Description: desc
        };
        const [year, month, day] = eatingDate.split("-");
        onChangeIsLoading(true);
        fetch(`${backendUrl}/records/${year}/${month}/${day}`, {
            method: 'PUT',
            headers: HEADERS,
            body: JSON.stringify(newRecord)
        })
            .then((resp) => resp.json())
            .then((_) => {
                const d = thisMonthRecord;
                d[newRecord.EatingDate] = newRecord;
                d[newRecord.EatingDate].isModifying = true;
                onChangeThisMonthRecord(d);
                onChangeIsLoading(false);
                closeForm();
            }).catch((error) => {
                Alert.alert("Error", error.message, [{ text: "Okay" }]);
                onChangeIsLoading(false);
            });
    }

    return (
        <View>
            <Text>Record today's food</Text>
            <Text>Food name</Text>
            {/* TODO: add auto complete feature from database */}
            <TextInput
                onChangeText={onChangeName}
                value={name}
                defaultValue={name}
                placeholder="Enter the food name"
                autoFocus={true}
            />
            <Text>Satisfaction Score</Text>
            <SatisfactionScoreDropdown initScore={score} handle={onChangeScore} />
            <Text>Eating Date</Text>
            <Text>{eatingDate}</Text>
            <Text>Eaten quantity</Text>
            <TextInput
                onChangeText={onChangeQuantity}
                value={quantity}
                defaultValue={quantity.toString()}
                placeholder="How many bags or cans?"
                keyboardType="numeric"
            />
            <Text>Food description</Text>
            <TextInput
                onChangeText={onChangeDesc}
                value={desc}
                defaultValue={desc}
                placeholder="What's about the food?"
            />
            <Button
                title={record.isModifying ? "Update" : "Add"}
                onPress={record.isModifying ? updateRecord : submitRecord}
                disabled={isLoading}
            />
            <Button title="Cancel" onPress={closeForm} />
        </View>
    )
}
