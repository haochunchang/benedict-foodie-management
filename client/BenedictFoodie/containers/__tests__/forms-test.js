import React from 'react';
import renderer from 'react-test-renderer';
import { FoodForm, RecordForm } from '../forms';

test('renders FoodForm correctly', () => {
    const tree = renderer.create(<FoodForm />).toJSON();
    expect(tree).toMatchSnapshot();
});

test('renders RecordForm correctly', () => {
    const tree = renderer.create(<RecordForm />).toJSON();
    expect(tree).toMatchSnapshot();
});