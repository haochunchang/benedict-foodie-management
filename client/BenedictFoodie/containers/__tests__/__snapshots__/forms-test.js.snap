// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FoodForm correctly 1`] = `
<View>
  <Text>
    Creat Food Stock
  </Text>
  <Text>
    Food name
  </Text>
  <TextInput
    autoFocus={true}
    onChangeText={[Function]}
    placeholder="Enter the food name"
    value=""
  />
  <Text>
    Food type
  </Text>
  <View
    style={
      Array [
        Object {
          "alignSelf": "stretch",
        },
        Object {
          "paddingBottom": 0,
          "zIndex": 999,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 2,
            "flexDirection": "row",
            "height": 40,
            "justifyContent": "space-between",
          },
          undefined,
          Object {
            "paddingHorizontal": 10,
          },
        ]
      }
    >
      <Text
        lineBreakMode="tail"
        numberOfLines={1}
        style={
          Array [
            undefined,
            Object {
              "width": 187,
            },
          ]
        }
      >
        Dry Food
      </Text>
    </View>
  </View>
  <Text>
    Purchase Date
  </Text>
  <TextInput
    keyboardType="numeric"
    onChangeText={[Function]}
    placeholder="Enter the purchase date in YYYY-MM-DD format"
    value="2022-10-29"
  />
  <Text>
    Food quantity
  </Text>
  <TextInput
    keyboardType="phone-pad"
    onChangeText={[Function]}
    placeholder="Enter the number of bags or cans"
    value={0}
  />
  <Text>
    Food description
  </Text>
  <TextInput
    onChangeText={[Function]}
    placeholder="What's about the food?"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Submit
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Cancel
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders RecordForm correctly 1`] = `
<View>
  <Text>
    Record today's food
  </Text>
  <Text>
    Food name
  </Text>
  <TextInput
    autoFocus={true}
    defaultValue=""
    onChangeText={[Function]}
    placeholder="Enter the food name"
    value=""
  />
  <Text>
    Satisfaction Score
  </Text>
  <View
    style={
      Array [
        Object {
          "alignSelf": "stretch",
        },
        Object {
          "paddingBottom": 0,
          "zIndex": 999,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "backgroundColor": "#efefef",
            "borderRadius": 2,
            "flexDirection": "row",
            "height": 40,
            "justifyContent": "space-between",
          },
          undefined,
          Object {
            "paddingHorizontal": 10,
          },
        ]
      }
    >
      <Text
        lineBreakMode="tail"
        numberOfLines={1}
        style={
          Array [
            undefined,
            Object {
              "width": 187,
            },
          ]
        }
      >
        Does not eat with snack added
      </Text>
    </View>
  </View>
  <Text>
    Eating Date
  </Text>
  <TextInput
    defaultValue=""
    keyboardType="numeric"
    onChangeText={[Function]}
    value=""
  />
  <Text>
    Eaten quantity
  </Text>
  <TextInput
    defaultValue="0"
    keyboardType="numeric"
    onChangeText={[Function]}
    placeholder="How many bags or cans?"
    value="0"
  />
  <Text>
    Food description
  </Text>
  <TextInput
    defaultValue=""
    onChangeText={[Function]}
    placeholder="What's about the food?"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Add
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Cancel
      </Text>
    </View>
  </View>
</View>
`;
